### PR TITLE
Add CSRF token for Ajax

### PR DESCRIPTION
**Pull request by asika32764**
### Summary of Changes

An idea inspired by Laravel when I'm writing drag and drop installation in https://github.com/joomla/joomla-cms/pull/14924

Add CSRF token to head as a meta tag, so we can get the value by Javascript.

Use `JHtml::_('form.csrf')` to add `<meta name="csrf-token" content="14044d42e0488e8362c2ee40f3dcf2d1" />` in `<head>`

Then we can select this tag to get value:

``` js
var token = jQuery('meta[name=csrf-token]').val();
```

Also we can use `X-Csrf-Token: xxx` header to validate token.

`JHtml_('jquery.token')` help us add header to `ajaxSetup` so we can auto inject token in every `$.ajax()` calls.

``` js
;(function ($) {
	$.ajaxSetup({
		headers: {
			'X-Csrf-Token': $('meta[name="csrf-token"]').attr('content')
		}
	});
})(jQuery);
```

### Testing Instructions

I rewrite sendmail test in configuration to use this pattern, just send a test.

global_configuration_-_joomladev_-_administration

If we remove `JHtml::_('jquery.token')` in `administrator/components/com_config/view/application/tmpl/default_mail.php`, will get fail result.

global_configuration_-_joomladev_-_administration2

### CODE REVIEW

File: libraries/joomla/session/session.php [hunk 1]
```diff
@@ -311,6 +311,20 @@ public static function checkToken($method = 'post')
 		$token = self::getFormToken();
 		$app = JFactory::getApplication();
 
+		// Check from header first
+		$csrf = isset($_SERVER['HTTP_X_CSRF_TOKEN']) ? $_SERVER['HTTP_X_CSRF_TOKEN'] : null;
```
Comment (mbabker) [inline libraries/joomla/session/session.php]: `$app->input->server->get()` works here.

**Comment by dgt41** [comment]
@asika32764 can you add the post header also to core Joomla.request ?
https://github.com/joomla/joomla-cms/blob/staging/media/system/js/core-uncompressed.js#L792-L888

**Comment by asika32764** [comment]
OK, but what class and method should we locate the php code? `JHtmlBehavior` ???

**Comment by dgt41** [comment]
`JHtmlForm::csrf()`is fine, you already oded that part

**Comment by asika32764** [comment]
I mean, the `core.js` is included by `JHtmlBehavior::core()` but it has no not added CSRF token yet.

Should I call `JHtmlForm::csrf()` also in `JHtmlBehavior::core()` or add a new method `JHtmlBehavior::token()` similar as `JHtmlJquery::token()` which I wrote in this PR?

**Comment by dgt41** [comment]
Personally I would extend the `JHtmlBehavior::core()` so everything comes up magically...

**Comment by asika32764** [comment]
OK. I just think there may be someone want to use core.js but do not hope CSRF auto inject to `<head>`.

But I will follow your comment first if there has no opinion from others.

**Comment by asika32764** [comment]
Done. Please test with this code in console:

``` js
Joomla.request({
  url: '.',
  method: 'POST',
  data: {foo: 'bar'}
});
```

articles_-_joomladev_-_administration

I'm wonder our `request` method do not support `PUT`, `PATCH` and `DELETE`???

File: media/system/js/core-uncompressed.js [hunk 2]
```diff
@@ -829,19 +829,26 @@ Joomla.editors.instances = Joomla.editors.instances || {
...
 
 			// Set the headers
 			xhr.setRequestHeader('X-Requested-With', 'XMLHttpRequest');
 			xhr.setRequestHeader('X-Ajax-Engine', 'Joomla!');
 
-			if (options.method === 'POST' && (!options.headers || !options.headers['Content-Type'])) {
-				xhr.setRequestHeader('Content-Type', 'application/x-www-form-urlencoded');
+			if (options.method === 'POST') {
+				if (document.querySelector('meta[name="csrf-token"]').getAttribute('content')) {
+					xhr.setRequestHeader('X-Csrf-Token', document.querySelector('meta[name="csrf-token"]').getAttribute('content'));
```
Comment (brianteeman) [inline media/system/js/core-uncompressed.js]: I am not 100% certain but from what I can see this should be 
'X-CSRF-Token'

Note the capitalisation

**Comment by dgt41** [comment]
> I'm wonder that our request method does not support PUT, PATCH and DELETE???

Maybe forcing this behaviour only to POST is not a great idea...

File: media/system/js/core-uncompressed.js (hunk 2 above)
Comment (asika32764) [reply media/system/js/core-uncompressed.js]: Yes you're right, since header is case insensitive but It is often use `CSRF` not `Csrf` in header.

**Comment by asika32764** [comment]
> Maybe forcing this behaviour only to POST is not a great idea...

I think we can leave this issue for J4 since JInput in J3 is not good at handling pure PUT, PATCH and DELETE data.

**Comment by mbabker** [comment]
JInput handles them just fine.  The magic getter in JInput looks for superglobals or subclasses, and there isn't something matching that definition for any request method except GET and POST.

**Comment by asika32764** [comment]
Maybe I'm miss something but to my knowledge, PUT request data must read by `fopen('php://input')`,  it cannot get from `$_PUT`. But JInput didn't handle it. (I only see `JInputJSON` did it)

**Comment by mbabker** [comment]
Right, so we already cover that case.  GET and POST requests have corresponding superglobals, the other request types don't, and I don't think we should be changing JInput's magic getter to proxy a request method into another object instance personally.

The thing that falls over is if you blindly try to do `$input->$method->get()` without validating `$method` (which, sadly, core actually does this).  IMO it's not hard to do something like this:

```php

$input = JFactory::getApplication()->input;
$method = $input->getMethod();

if ($method === 'PUT') {
    $data = $input->json->get('foo');
elseif (in_array($method, ['GET', 'POST'])) {
    $data = $input->$method->get('foo');
} else {
    throw new InvalidArgumentException(sprintf('The "%s" request method is not supported by this route, please send a GET, POST, or PUT request.', $method), 405);
}
```

**Comment by asika32764** [comment]
Actually I think `$input->put->get()` is a nice pattern since the main subject of `JInput` is to help us integrate all system INPUT in one interface. 

Maybe add a `JinputFormdata` to parse `php://input` instead get value from `$_REQUEST`. [This is what I'm working in my framework](https://github.com/ventoviro/windwalker/blob/master/src/IO/FormDataInput.php#L96-L113)

But I think it is too big subject for this PR, let's leave this issue for the Framework design, not discussion here.

File: libraries/cms/html/form.php [hunk 3]
```diff
@@ -17,17 +19,62 @@
...
+
+		/** @var JDocumentHtml $doc */
+		$doc = JFactory::getDocument();
+
+		if ($doc->getType() !== 'html' || !$doc instanceof JDocumentHtml)
+		{
+			return;
+		}
+
+		$doc->setMetaData($name, JSession::getFormToken());
```
Comment (Fedik) [inline libraries/cms/html/form.php]: I think, here we could use the "script options" feature, without extra metatags, and so without extra DOM query,
see [Adding_the_options_to_your_JavaScript_code](https://docs.joomla.org/J3.x:Adding_JavaScript_and_CSS_to_the_page#Adding_the_options_to_your_JavaScript_code)

File: media/system/js/core-uncompressed.js [hunk 4]
```diff
@@ -829,24 +829,31 @@ Joomla.editors.instances = Joomla.editors.instances || {
...
 			xhr.open(options.method, options.url, true);
 
 			// Set the headers
 			xhr.setRequestHeader('X-Requested-With', 'XMLHttpRequest');
 			xhr.setRequestHeader('X-Ajax-Engine', 'Joomla!');
 
-			if (options.method === 'POST' && (!options.headers || !options.headers['Content-Type'])) {
-				xhr.setRequestHeader('Content-Type', 'application/x-www-form-urlencoded');
+			if (options.method === 'POST') {
+				if (document.querySelector('meta[name="csrf-token"]').getAttribute('content')) {
```
Comment (Fedik) [inline media/system/js/core-uncompressed.js]: `document.querySelector('meta[name="csrf-token"]')` return `null` when nothing found, and whole code fail

File: libraries/cms/html/form.php (hunk 3 above)
Comment (asika32764) [reply libraries/cms/html/form.php]: It's seems better then Laravel, I will try implement it later.

**Comment by asika32764** [comment]
Here I have a security concern, if we auto inject CSRF into all request, can there be a case that developer or 3rd extensions send a request with CSRF token to outside of Joomla without aware?

**Comment by asika32764** [comment]
Use script option to handle CSRF

File: libraries/cms/html/form.php [hunk 5]
```diff
@@ -17,17 +19,62 @@
...
 	 *
+	 * @param   array  $attribs  Input element attributes.
+	 *
 	 * @return  string  A hidden input field with a token
 	 *
 	 * @see     JSession::checkToken()
 	 * @since   1.5
 	 */
-	public static function token()
+	public static function token($attribs = null)
```
Comment (mbabker) [inline libraries/cms/html/form.php]: `token(array $attribs = array()`

The default value might as well be the correct data type.

File: libraries/cms/html/form.php [hunk 6]
```diff
@@ -17,17 +19,62 @@
...
+	{
+		if (is_array($attribs))
```
Comment (mbabker) [inline libraries/cms/html/form.php]: Then this can become a `!empty($attribs)` check

File: libraries/cms/html/form.php [hunk 7]
```diff
@@ -17,17 +19,62 @@
...
+		{
+			$attribs = ' ' . ArrayHelper::toString($attribs);
+		}
+
+		return '<input type="hidden" name="' . JSession::getFormToken() . '" value="1"' . $attribs . ' />';
+	}
+
+	/**
+	 * Add CSRF form token to <head> meta that developers can get it by Javascript.
```
Comment (mbabker) [inline libraries/cms/html/form.php]: Since this isn't adding a meta tag anymore the summary can be updated.

**Review by mbabker — CHANGES_REQUESTED** [review]

**Review by mbabker — APPROVED** [review]

**Comment by zero-24** [comment]
I have tested this item :white_check_mark: successfully on f9e0372723112010510dd52458c3b1f924e678c1<br /><br />After installing a new site based on that branch here the send mail button still works. Thanks<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14952">issues.joomla.org/tracker/joomla-cms/14952</a>.</sub>

**Comment by ghost** [comment]
@zero-24 can this be local tested (mamp) using Mailer "Sendmail"?

**Comment by zero-24** [comment]
I don't know about mamp but in xampp there is a folder where all mails get saved. Just check how it works before the patch. As it should work like the same after the patch

**Comment by ghost** [comment]
Without Patch using Default-Path "/usr/sbin/sendmail" get after Click on Button "Send Test Mail" `Error Test mail could not be send` and `Notice Could not execute: /usr/sbin/sendmail`.

If i get same Result using PR is this a successfully Test?

**Comment by zero-24** [comment]
hmm IMO it would better to do a working test.

**Comment by ghost** [comment]
thanks, hopefully another one will test.

**Comment by asika32764** [comment]
If you are using MAMP in MacOS, I guess select the `PHP Mail` option may work.

**Comment by ghost** [comment]
@asika32764 none of the 3 works, this is from "PHP Mail":
bildschirmfoto 2017-07-29 um 13 47 55

**Comment by zero-24** [comment]
@asika32764 Isn't the above error coming from the local envoirment? As this should have nothing todo with the patch here?

**Comment by ghost** [comment]
i haven't applied the PR, so this Notice and Errors don't come from PR.

**Comment by asika32764** [comment]
I think

**Comment by asika32764** [comment]
A mistake that deleted the branch from forked repo, just restored it.

**Comment by asika32764** [comment]
@franz-wohlkoenig Actually I think it is successfully test if it return same message before and after patch.

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on f9e0372723112010510dd52458c3b1f924e678c1<br /><br />Got same Messages with PR as without.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14952">issues.joomla.org/tracker/joomla-cms/14952</a>.</sub>

**Comment by ghost** [comment]
RTC after two successful tests.

**Comment by pawan0446** [comment]
I have not tested this item.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14952">issues.joomla.org/tracker/joomla-cms/14952</a>.</sub>

**Pull request merged**